File: model/bardis_model.r
# Define the system of differential equations
bardis_model <- function(time, state, parameters) {
  with(as.list(c(state, parameters)), {
    
    # Flows

    #Resources 
    resource_inflow <- k_resources * Resources
    extractive_pollution <- k_pollution * Economy * Resources 

    #Economy 
    production <- ef_economy_resources_on_prod * Resources * Economy + ef_bureaucracy_on_prod * Bureaucracy  
    depreciation <- k_deprec * Economy + ef_pollution_on_depreciation * Pollution 
    
    #Bureaucracy 
    bureaucracy_creation <- ef_economy_on_bureaucracy * Economy + k_bureaucracy * Bureaucracy
    bureaucracy_decay <- k_decay_bureaucracy  * Bureaucracy + ef_pollution_on_bureaucracy * Pollution 

    #Pollution 
    pollution_abtement <- k_pollution_decay * Pollution

    # Rate of change
    dResources <- ifelse(Resources > 0, resource_inflow - production - extractive_pollution, 0)
    dEconomy <- ifelse(Economy > 0, production - depreciation - bureaucracy_creation, 0)
    dBureaucracy <- ifelse(Bureaucracy > 0, bureaucracy_creation - bureaucracy_decay, 0)
    dPollution <- ifelse(Pollution > 0, depreciation + bureaucracy_decay + extractive_pollution - pollution_abtement, 0)   
    list(c(dResources, dEconomy, dBureaucracy, dPollution))
  })
}

File: model/python_ver/bardis_model.py
class BardisModel:

    def __init__(self, parameters):
        self.parameters = parameters

    # TODO: Complete the class



def bardis_model(state, time, parameters):
    # Unpack state variables
    Resources, Economy, Bureaucracy, Pollution = state

    # Unpack parameters
    k_resources = parameters["k_resources"]
    k_pollution = parameters["k_pollution"]
    ef_economy_resources_on_prod = parameters["ef_economy_resources_on_prod"]
    ef_bureaucracy_on_prod = parameters["ef_bureaucracy_on_prod"]
    k_deprec = parameters["k_deprec"]
    ef_pollution_on_depreciation = parameters["ef_pollution_on_depreciation"]
    ef_economy_on_bureaucracy = parameters["ef_economy_on_bureaucracy"]
    k_bureaucracy = parameters["k_bureaucracy"]
    k_decay_bureaucracy = parameters["k_decay_bureaucracy"]
    ef_pollution_on_bureaucracy = parameters["ef_pollution_on_bureaucracy"]
    k_pollution_decay = parameters["k_pollution_decay"]

    # Compute flows
    resource_inflow = k_resources * Resources
    extractive_pollution = k_pollution * Economy * Resources

    production = ef_economy_resources_on_prod * Resources * Economy + ef_bureaucracy_on_prod * Bureaucracy
    depreciation = k_deprec * Economy + ef_pollution_on_depreciation * Pollution

    bureaucracy_creation = ef_economy_on_bureaucracy * Economy + k_bureaucracy * Bureaucracy
    bureaucracy_decay = k_decay_bureaucracy * Bureaucracy + ef_pollution_on_bureaucracy * Pollution

    pollution_abatement = k_pollution_decay * Pollution

    # Compute derivatives with conditional checks
    dResources = resource_inflow - production - extractive_pollution if Resources > 0 else 0
    dEconomy = production - depreciation - bureaucracy_creation if Economy > 0 else 0
    dBureaucracy = bureaucracy_creation - bureaucracy_decay if Bureaucracy > 0 else 0
    dPollution = depreciation + bureaucracy_decay + extractive_pollution - pollution_abatement if Pollution > 0 else 0

    return [dResources, dEconomy, dBureaucracy, dPollution]
